pkg/proxy/proxy.go

```diff
@@ -391,18 +392,52 @@ func (p *Proxy) addUserContext() echo.MiddlewareFunc {
 				return next(ctx)
 			}
 
-			userID, username, err := p.extractUserID(ctx.Request())
+			token, err := p.extractUserToken(ctx.Request())
 			if err != nil {
 				return crterrors.NewUnauthorizedError("invalid bearer token", err.Error())
 			}
-			ctx.Set(context.SubKey, userID)
-			ctx.Set(context.UsernameKey, username)
+			ctx.Set(context.SubKey, token.Subject)
+			ctx.Set(context.UsernameKey, token.PreferredUsername)
+			ctx.Set(context.EmailKey, token.Email)
 
 			return next(ctx)
 		}
 	}
 }
 
+// ensureUserIsNotBanned rejects the request if the user is banned
+func (p *Proxy) ensureUserIsNotBanned() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(ctx echo.Context) error {
+			errorResponse := func(err *crterrors.Error) error {
+				ctx.Logger().Error(errs.Wrap(err, "workspace list error"))
+				ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+				ctx.Response().Writer.WriteHeader(int(err.Code))
+				return json.NewEncoder(ctx.Response().Writer).Encode(err.Status)
+			}
+
+			email := ctx.Get(context.EmailKey).(string)
+			if email == "" {
+				return errorResponse(crterrors.NewUnauthorizedError("unauthenticated request", "anonymous access is not allowed"))
+			}
+
+			// retrieve banned users
+			uu, err := p.app.InformerService().BannedUsersByEmail(email)
+			if err != nil {
+				return errorResponse(crterrors.NewInternalError(errs.New("unable to retrieve user"), "could not define ban status"))
+			}
+
+			// if a matching Banned user is found, then user is banned
+			if len(uu) > 0 {
+				return errorResponse(crterrors.NewForbiddenError("user is banned", "user is banned"))
+			}
+
+			// user is not banned
+			return next(ctx)
+		}
+	}
+}
```

Comment:
We could probably add some unit tests for `ensureUserIsNotBanned()` if it's not too tricky, right? Unless you believe it's already indirectly covered somehow...

Reply:
done in b822ccf2581c0429e5fb37644e566577702f138b